lib/internal/webstreams/readablestream.js

```diff
@@ -658,11 +658,13 @@ class ReadableStreamBYOBRequest {
     const viewBuffer = ArrayBufferViewGetBuffer(view);
     const viewBufferByteLength = ArrayBufferGetByteLength(viewBuffer);
 
-    if (viewByteLength === 0 || viewBufferByteLength === 0) {
-      throw new ERR_INVALID_STATE.TypeError(
-        'View ArrayBuffer is zero-length or detached');
+    if (isDetachedBuffer(viewBuffer)) {
+      throw new ERR_INVALID_STATE.TypeError('Viewed ArrayBuffer is detached');
     }
 
+    assert(viewByteLength > 0);
+    assert(viewBufferByteLength > 0);
```

Comment:
This probably requires a comment, e.g. referencing WPT tests that depend on it; or an explicit test. Otherwise it's tempting to be optimized away in some way.

Reply:
Well.. these are conditions we're assuming at this stage. If any tests fail because of these, that means our implementation is something missing or wrong. It looks enough without a comment.

Reply:
The question is: would a test fail if someone will undo changes in `respond()`?

> 2. If ! IsDetachedBuffer(this.[[view]].[[ArrayBuffer]]) is true, throw a TypeError exception.
> 3. Assert: this.[[view]].[[ByteLength]] > 0.
> 4. Assert: this.[[view]].[[ViewedArrayBuffer]].[[ByteLength]] > 0.

Implementation-wise, if we are guaranteed that `byteLength` can become zero only if the underlying `ArrayBuffer` is detached, these steps are effectively handled by old code.

Specs-wise, I read "Assert" here as "the following condition is guaranteed to be truth" rather than "if the following condition is false, throw".
If my interpretation is wrong, the new code is undoubtedly more spec-compliant, and needs a test that will expect `AssertionError` instead of `TypeError` (does wpt include such test?). If this condition can not be reliably simulated by test, a comment would be enough.
This is not a strong opinion; if you're sure that it's better without a comment, it can be left as is.

Reply:
> The question is: would a test fail if someone will undo changes in respond()?
> Implementation-wise, if we are guaranteed that byteLength can become zero only if the underlying ArrayBuffer is detached, these steps are effectively handled by old code.

Agree that we can handle the steps effectively if that's guaranteed. That sounds to me like we can use the old code to check for detaching also. However, it doesn't seem to work. The following is an example:
```js
function isDetachedBuffer(buffer) {
   try {
      new Uint8Array(buffer);
    } catch {
      return true;
    }
  return false;
}

const view = new Uint8Array();
assert(isDetachedBuffer(view.buffer) == false); // This should be false although byteLength is 0.
```


> a test that will expect `AssertionError` instead of `TypeError` (does wpt include such test?).

AFAK, wpt doesn't include tests to verify such assertions work. I'd like to follow how to handle the other `assert`s in this file.

Reply:
> However, it doesn't seem to work.

Not really sure what you meant here. Detached `ArrayBuffer` is indeed just a subset of zero-length ones in general, but it might be the only possible reason in the particular contexts, where normal zero-length buffers are filtered out. `byteLength` is immutable in normal circumstances, AFAIK it can be resetted only by detaching.

Overall the main idea is that `ArrayBuffer` with zero length is a pretty useless entity, so in most usecases we want to immediately return or throw anyway. Aside from spec compliance and situations where we really want to inform user about detaching, `byteLength` check will always look like a good optimization.

Reply:
> Aside from spec compliance and situations where we really want to inform user about detaching, byteLength check will always look like a good optimization.

Agreed. It's worth considering optimizing `IsDetachedBuffer` in a way to check only `byteLength` as a trade-off for the performance. Although it may cause losing compatibility in a few non-practical use cases (A few of wpt would fail.), but get better performance in most practical use cases.

We could do such a performance optimization including the way of C++ builtin in a follow-up since Web Streams API is still experimental.